vignettes/rspm-snaphots.Rmd

```diff
@@ -0,0 +1,223 @@
+This practice has the following advantages in our view:
+
+-   Stable project environment by making use of packages which are likely to play together well.
+-   Package versions which are known to work with the R version used.
+-   Possibility of controlled package updates between RSPM snapshots.
+
+The combination of an specific R version and a RSPM snapshot is ensured by {renv}.
+In `renv.lock`, the R version and snapshot ID are listed.
+
+``` {.json}
+{
+  "R": {
+    "Version": "4.1.0",
+    "Repositories": [
+      {
+        "Name": "CRAN",
+        "URL": "https://packagemanager.rstudio.com/cran/2021-05-18"
+      }
+    ]
+  },
+  [...]
+}
+```
+
+## The nature of snapshot-centered workflows
+
+Now when a project is initiated with {renv} and packages are installed, only packages from this specific snapshot are installed.
+A "snapshot" here means that the repo where packages are looked up contains only the CRAN sources of this specific day in time.
+No additional packages will be available/added to this snapshot in the future.
+
+This might seem limiting on the first look but we believe that it is actually a feature.
+Often when collaborating with other people, multiple people install packages at different stages of a project lifetime.
+
+Let's compare the scenario of a `install.packages("dplyr")` call eight weeks after project start.
+**Without a fixed snapshot**, this would look as follows:
+
+1.  A user installs {dplyr} five weeks after the project has started.
+2.  Meanwhile, {dplyr} was updated and with it some dependencies of it (maybe even the minimum versions were bumped). Hence, `install.packages("dplyr")` will also update more recent versions of other packages than just {dplyr}.
+3.  These updates have the potential to break some analysis in the project without anyone realizing.
+
+In contrast, when **using a snapshot-centered approach**, `install.packages("dplyr")` eight weeks after project start
+
+1.  will not cause any updates of already installed packages bcausse the {dplyr} version which is going to be installed will be the {dplyr} version at the point in time referenced by the RSPM snapshot.
+2.  Already installed packages will not be installed again (thanks to {renv} and most importantly, no updates will be installed.
+
+## Practical implications and usage
+
+### Using packages not available in the RSPM snapshot
+
+Often enough there is the need to use packages which are not available in the RSPM snapshot, for example packages living only on GitHub or newer versions of packages.
+In general we recommend trying to get along with the packages available in the snapshot.
+If this is not possible for whatever reason, there is always the possibility to install a specific package from GitHub and track its exact version via {renv}.
```

Comment:
These days I tend to bump to a new snapshot but avoid upgrading all packages, only upgrading those that are necessary to achieve a feature. A lifting of all packages can be done separately shortly after that. 

Installing from GitHub won't give us binaries, unfortunately.

Reply:
Hm we might want to get this "right" to avoid confusion for clients.

If I tell client we do it like X and then sometimes do it like Y, this raises confusion.

I see the reasons for the partly bump but I think this might bring in confusion.
Do we want to follow a more "clear" approach here?

I.e. we could say to only lift full snapshots / introduce new ones if more than 3 (?) packages need to be lifted (for whatever reasons).
Otherwise we install the needed packages from GH.
Installing two packages or so from GH is fine I'd argue and the footprint is less than bumping the whole snapshot?

I think if we bump a snapshot, we should lift all packages and not only some.
The reasoning is that if someone installs a new package, this installation might usually update multiple packages.

Also because I'd like to promote the general approach to the community, I'd like it to be "clear" in this relation. Without full snapshot bumps we are again somewhere between a snapshot-centered approach and a rolling CRAN one because `install.packages()` would bring in new packages.

Reply:
I think installing from GH if the version isn't available in the snapshot adds more confusion.

Bumping snapshots can be independent from lifting, this has worked well in the past. Changing the installed packages is a conscious exercise.

How about recommending lifting "soon" after bumping the snapshot date?

Reply:
One reason to advocate for independence: lifting touches the whole `renv.lock` file, updating individual packages touches only a few entries.

Reply:
It's too much of a hassle to always upgrade everything if we need updates from a single package. Most of the time updates tend to work well together, we protect against the very rare case of incompatibilities/inconsistencies -- and then there's always the option of lifting or even selectively downgrading.

Reply:
Can we add a flow diagram?

Reply:
What happens if one does an `renv::restore()` when the snapshot was lifted and only some pkgs were updated? `renv::restore()` will try to find the package versions within the new snapshot and will fail, won't it?

> How about recommending lifting "soon" after bumping the snapshot date?

I'd really like to avoid such recommendations for clients since they might get unsure about what to do if things are not fully clear; ala "what is soon?"
I see the point and don't get me wrong, would have no problems with this flexible approach in general - but I think the more flexible we make it, the more confusion and unclear states we'll get in a project.

> One reason to advocate for independence: lifting touches the whole renv.lock file, updating individual packages touches only a few entries.

The "partly lifting" approach looks more non-snapshot centered to me when looking at the big picture.
Let's sketch the following scenario: Given that one joins a project which runs on an RSPM snapshot - how does one see which packages are installed from which snapshot date with this flexible approach? This information is not stored individually in each package entry.

At first I would assume all packages refer to the snapshot date in `renv.lock` but if the snapshot was manually lifted and only some packages were updated, there is no reference anymore which packages are from which snapshot.

This might cause potential trouble then and I see project states which have like 30% updated packages and 70% "old" packages.

The more I think about it, the more I favor a "full lift" or "no lift". Or even snapshot + GH > partly lifting packages via a snapshot bump.

Reply:
`restore()` always works and always restores the correct package versions in this scenario.

I believe that the recommendation to always lift everything causes unnecessary friction. Bumping the snapshot and individual packages is a good compromise.

Will this flow diagram (or maybe several) help in reasoning about the individual steps and the big picture? I paste a rendering in the comments below.

```mmd
graph TD
  cynkrathis_init["cynkrathis::init_renv()"] --> happy
  init["init()"] --> snapshot0["snapshot()"]
  snapshot0["snapshot()"] --> happy
  happy --> restore0["restore()"]
  restore0 --> happy["(((happy)))"]
  happy --> work
  work --> install0["install() | update()"]
  install0 --> work
  work --> restore["restore(clean = TRUE)"]
  restore --> install["install() | update()"]
  install --> snapshot["snapshot()"]
  install -.-> work
  snapshot --> happy
  happy --> conserve
  conserve --> restore
  restore -.-> happy
```

Reply:
> restore() always works and always restores the correct package versions in this scenario.

I assume it gets a package not available in `repos` from the cache?

Will this work if a collaborator does not have access to the cache? I.e. one clones the repo and then calls `restore()`?
In this scenario I assume that `renv` would

- look in the cache -> no hit
- look in the `repos` -> no hit

-> fail ultimatively?

> Will this flow diagram (or maybe several) help in reasoning about the individual steps and the big picture? I paste a rendering in the comments below.

Thanks, will this render in an Rmd?

Reply:
I think it's important to ensure that packages that are mentioned in the lock file are installable from the repo. We bump the snapshot link in `renv.lock`, restart the R session, and install the package in the correct version. After that we don't change the snapshot link back, this has to be one unit.

Mermaid can be rendered in `.Rmd` with {DiagrammeR}. I'll dig up an example when necessary.

Reply:
> I think it's important to ensure that packages that are mentioned in the lock file are installable from the repo.

Sure, but is this true when following the "partial bump" approach? Some packages won't correspond to the snapshot ID in `renv.lock` anymore and hence not downloadable in a vanilla environment.

Hence, in my example above I still think that new people cloning the project would potentially fail to `restore()`?
If so, this would be a blocker for the "partial bump" approach, wouldn't it?